src/gtt_journal.c

```diff
@@ -829,23 +859,58 @@ html_url_requested_cb (GtkHTML *doc, const gchar *url, GtkHTMLStream *handle,
   if (!path)
     return;
 
-  GnomeVFSResult result;
-  GnomeVFSHandle *vfs;
-  result = gnome_vfs_open (&vfs, path, GNOME_VFS_OPEN_READ);
+  GFile *ifile = g_file_new_for_path (path);
 
-  if (GNOME_VFS_OK != result)
-    return;
+  GError *error = NULL;
+  GFileInputStream *istream = g_file_read (ifile, NULL, &error);
+  if (NULL == istream)
+    {
+      g_warning ("Failed to open file for reading: %s", error->message);
+
+      g_error_free (error);
+      error = NULL;
+
+      g_object_unref (ifile);
+      ifile = NULL;
+
+      return;
+    }
 
 #define BSZ 16000
   char buff[BSZ];
-  GnomeVFSFileSize bytes_read;
-  result = gnome_vfs_read (vfs, buff, BSZ, &bytes_read);
-  while (GNOME_VFS_OK == result)
+  while (1)
     {
+      const gsize bytes_read = g_input_stream_read (G_INPUT_STREAM (istream),
+                                                    buff, BSZ, NULL, &error);
+      if (0 > bytes_read)
+        {
+          g_warning ("Failed to read from file: %s", error->message);
+
+          g_error_free (error);
+          error = NULL;
+
+          break;
```

Comment:
You seem to be doing a lot of cut-n-paste of exactly the same code, making things verbose. You could just write this code once, stick it in a subroutine, call that subroutine.  It would then be a one-or-two lines of code change: replace `gnome_vfs_read` by `gtt_gio_read` and then do all of the `g_warning` and `g_error_free stuff inside of `gtt_gio_free`. This makes the code smaller and more compact, and easier to review.

It also makes the code "future proof" -- in the future, if you ever have to change the `g_warning` for some reason, then it would have to be changed in only one location, and not a dozen different ones. This avoids accidentally missing one of the dozen locations, or otherwise making mistakes in those kinds of changes.

Reply:
I wonder if in some of these cases the error object should ideally instead be received from the caller, and the caller would be responsible for display-and-free, instead of just dumping it as a g_warning and then not telling the caller that an error occurred. Like exception bubbling. That would probably reduce duplication a bit here.

Well, I'm not sure... and either way, it would probably be a too big refactoring at this stage of the modernization.

So yeah, if we are "hiding" errors deep down, I guess we could deduplicate like Lines suggests. Or even just gtt_warn_and_clear(&error) to at least collect those bits.

Reply:
Hi guys,

I myself am not happy with this code :-) If it were a greenfield implementation I would encapsulate much more logic. This is not my first try of refactoring GnoTime and one major reason for the failures of my previous attempts were my lack to resist premature optimization ;-)

The changes of this pull request are indeed bulky and clumsy, but my general aim is to stick to the existing structure of GnoTime as faithful as possible during the upgrade of the dependencies and only start refactoring and optimization when GnoTime is able to run natively in a modern environment without hacks.

Like @oskarb in a few cases I'm not feeling good with the handling and reporting of errors, but this is something I would leave for the second "post-dependency-upgrade" stage (see my suggestion about TODOs below).

As a suggestion of a compromise I will try to refactor this pull request in the next days to fix the most obvious code duplications and superfluous verbosity while still staying faithful to the current logic. Furthermore I would suggest that we start generously adding TODO markers throughout the code where we think something could be done better when GnoTime runs with modern dependencies.

Reply:
@oskarb I updated this pull request on the new HEAD. The code is still too verbose and not deduplicated. My preferred further path would be to get rid of the deprecated dependencies first while reproducing the behavior of `2.4.1` as close as possible. As soon as we are sure that "GnoTime 3" with modern dependencies has no regressions towards `2.4.1` I would suggest to conduct the refactoring before releasing `3.0.0` (whenever that may be ;-)).

Reply:
Thank you for the repeated review, I incorporated the suggestions and will merge now.